src/snowflake/snowpark/mock/file_operation.py

```diff
@@ -0,0 +1,178 @@
+_STAGE_FILE_MAP = defaultdict(set)
+
+PUT_RESULT_KEYS = [
+    "source",
+    "target",
+    "source_size",
+    "target_size",
+    "source_compression",
+    "target_compression",
+    "status",
+    "message",
+]
+
+SUPPORTED_CSV_READ_OPTIONS = (
+    "SKIP_HEADER",
+    "SKIP_BLANK_LINES",
+    "FIELD_DELIMITER",
+    "FIELD_OPTIONALLY_ENCLOSED_BY",
+)
+
+
+def put(local_file_name: str, stage_location: str) -> TableEmulator:
+    """
+    Put a file into in memory map, key being stage location and value being the local file path
+    """
+    local_file_name = local_file_name[
+        8:-1
+    ]  # skip normalized prefix `file:// and suffix `
+    file_name = os.path.basename(local_file_name)
+    remote_file_path = f"{stage_location}/{file_name}"
+    _FILE_STAGE_MAP[remote_file_path] = local_file_name
+    _STAGE_FILE_MAP[f"{stage_location}/"].add(local_file_name)
+    file_size = os.path.getsize(os.path.expanduser(local_file_name))
+
+    result_df = TableEmulator(
+        columns=PUT_RESULT_KEYS,
+        sf_types={
+            "source": StringType(),
+            "target": StringType(),
+            "source_size": DecimalType(10, 0),
+            "target_size": DecimalType(10, 0),
+            "source_compression": StringType(),
+            "target_compression": StringType(),
+            "status": StringType(),
+            "message": StringType(),
+        },
+        dtype=object,
+    )
+    result_df.loc[len(result_df)] = {
+        k: v
+        for k, v in zip(
+            PUT_RESULT_KEYS,
+            [file_name, file_name, file_size, file_size, None, None, None, None],
+        )
+    }
+    return result_df
+
+
+def read_file(
+    stage_location,
+    format: str,
+    schema: List[Attribute],
+    analyzer: "MockAnalyzer",
+    options: Dict[str, str],
+) -> TableEmulator:
+    try:
+        local_files = [_FILE_STAGE_MAP[stage_location]]
+    except KeyError:
+        if stage_location and stage_location[0] != "@":
+            raise SnowparkSQLException("SQL compilation error")
+        if stage_location in _STAGE_FILE_MAP.keys():
+            # all files within the stage
+            local_files = list(_STAGE_FILE_MAP[stage_location])
+        else:
+            raise
+    if format.lower() == "csv":
+        for option in options:
+            if option not in SUPPORTED_CSV_READ_OPTIONS:
+                _logger.warning(
+                    f"[Local Testing] read file option {option} is not supported."
+                )
+        skip_header = options.get("SKIP_HEADER", 0)
+        skip_blank_lines = options.get("SKIP_BLANK_LINES", False)
+        field_delimiter = options.get("FIELD_DELIMITER", ",")
+        field_optionally_enclosed_by = options.get("FIELD_OPTIONALLY_ENCLOSED_BY", None)
+        if (
+            field_delimiter[0]
+            and field_delimiter[-1] == "'"
+            and len(field_delimiter) >= 2
+        ):
+            field_delimiter = field_delimiter[1:-1]
```

Comment:
What are we trying to do here? I don't think I fully understand the if condition here.

Reply:
updated the comment, it's used to unquote the `field_delimiter`, we will have `"':'"`